prereise/gather/demanddata/bldg_electrification/puma_hp_estimator.py

```diff
@@ -0,0 +1,373 @@
+            "s_frac_elec_htg",
+            "s_log_hdd",
+            "s_natgas_price",
+        ],
+    )
+    fit_stats = pd.Series(
+        data=[
+            lm_hp.bse[0],
+            lm_hp.bse[1],
+            lm_hp.bse[2],
+            lm_hp.bse[3],
+            lm_hp.pvalues[1],
+            lm_hp.pvalues[2],
+            lm_hp.pvalues[3],
+            lm_hp.rsquared,
+        ],
+        index=[
+            "stderr_const",
+            "stderr_frac_elec_htg",
+            "stderr_log_hdd",
+            "stderr_natgas_price",
+            "pvalue_frac_elec_htg",
+            "pvalue_log_hdd",
+            "pvalue_natgas_price",
+            "rsquared",
+        ],
+    )
+    # It's observed that hp penetration has low variance when electric heating rate is low,
+    # and high variance hwne electric heating rate is high. So fit a simpler model with less
+    # variance for pumas with lower electric heating penetration as follows
+    lm_hp_low = sm.OLS(
+        stats["hp_penetration"],
+        sm.add_constant(
+            np.array(
+                stats[
+                    [
+                        "frac_elec_htg",
+                        "frac_elec_htg_sq",
+                    ]
+                ]
+            )
+        ),
+    ).fit()
+    hp_fit_df_low_elec = pd.Series(
+        data=[
+            lm_hp_low.params[0],
+            lm_hp_low.params[1],
+            lm_hp_low.params[2],
+        ],
+        index=[
+            "const",
+            "s_frac_elec_htg",
+            "s_frac_elec_htg_sq",
+        ],
+    )
+    fit_stats_low_elec = pd.Series(
+        data=[
+            lm_hp_low.bse[0],
+            lm_hp_low.bse[1],
+            lm_hp_low.bse[2],
+            lm_hp_low.pvalues[1],
+            lm_hp_low.pvalues[2],
+            lm_hp_low.rsquared,
+        ],
+        index=[
+            "stderr_const",
+            "stderr_frac_elec_htg",
+            "stderr_frac_elec_htg_sq",
+            "pvalue_frac_elec_htg",
+            "pvalue_frac_elec_htg_sq",
+            "rsquared",
+        ],
+    )
+
+    return hp_fit_df, fit_stats, hp_fit_df_low_elec, fit_stats_low_elec
+
+
+def estimate_puma_hp_penetration(hp_fit_df, hp_fit_df_low_elec, puma_data_metro):
+    """Compute current heat pump penetration for pumas, and then scale the values to align with regional inputs from
+         RECS and CBECS surveys
+
+    :param pandas.DataFrame hp_fit_df: slope and intercept of heat pump penetration regression model, output of
+        hp_penetration_fit()
+    :param pandas.DataFrame hp_fit_df_low_elec: slope and intercept of heat pump penetration regression model for
+        low electrified locations, output of hp_penetration_fit()
+    :param pandas.DataFrame puma_data_metro: puma data within metropolitan areas, output of puma_county_shp_overlay()
+
+    :return: (*pandas.DataFrame*) hp_puma_df: current heat pump penetration estimations for pumas that has been scaled
+        based on RECS and CBECS surveys
```

Comment:
Redundant newline. Let's add row and column info for the returned data frame.

Reply:
done